sdk/schemaregistry/schema-registry-avro/test/withMessagingClients.spec.ts

```diff
@@ -17,45 +30,136 @@ import { testGroup } from "./utils/dummies";
+  /**
+   * Each unit test correspond to one of the scenarios below
+   */
+  createScenario1Client: () => MessagingTestClient<T>;
+  createScenario2Client: () => MessagingTestClient<T>;
+  createScenario3Client: () => MessagingTestClient<T>;
+  createScenario4Client: () => MessagingTestClient<T>;
 }
 
-describe("With messaging clients", function () {
+describe.only("With messaging clients", function () {
   const eventHubsConnectionString = env.EVENTHUB_CONNECTION_STRING || "";
   const eventHubName = env.EVENTHUB_NAME || "";
+  const crossLanguage = env.CROSS_LANGUAGE !== undefined;
+
+  function createEventHubsTestClient(settings: { eventHubName: string; crossLanguage: boolean }) {
+    const { crossLanguage, eventHubName: inputEventHubName } = settings;
+    const client = createMockedMessagingClient(() =>
+      createEventHubsClient({
+        crossLanguage,
+        eventHubName: crossLanguage ? inputEventHubName : eventHubName,
+        eventHubsConnectionString,
+      })
+    );
+    client.initialize();
+    return client;
+  }
+
   const eventDataTestInfo: ScenariosTestInfo<EventData> = {
-    adapterFactory: createEventDataAdapter({
+    messageAdapter: createEventDataAdapter({
       properties: {
         language: "js",
       },
     }),
     messagingServiceName: "Event Hub",
-    client: createMockedMessagingClient(() =>
-      createEventHubsClient(eventHubsConnectionString, eventHubName)
-    ),
+    createScenario1Client: () =>
+      createEventHubsTestClient({
+        crossLanguage,
+        eventHubName: "scenario_1",
+      }),
+    createScenario2Client: () =>
+      createEventHubsTestClient({
+        crossLanguage,
+        eventHubName: "scenario_2",
+      }),
+    createScenario3Client: () =>
+      createEventHubsTestClient({
+        crossLanguage,
+        eventHubName: "scenario_3",
+      }),
+    createScenario4Client: () =>
+      createEventHubsTestClient({
+        crossLanguage,
+        eventHubName: "scenario_4",
+      }),
   };
   matrix([[eventDataTestInfo]] as const, async (testInfo: ScenariosTestInfo<any>) => {
-    describe(testInfo.messagingServiceName, async function () {
+    const {
+      messageAdapter,
+      messagingServiceName,
+      createScenario1Client,
+      createScenario2Client,
+      createScenario3Client,
+      createScenario4Client,
+    } = testInfo;
+    describe(messagingServiceName, async function () {
       let serializer: AvroSerializer<any>;
+
+      async function roundtrip(settings: {
+        client: MessagingTestClient<any>;
+        value: unknown;
+        writerSchema: string;
+        processMessage: (p: Promise<unknown>) => Promise<void>;
+        readerSchema?: string;
+        eventCount?: number;
+      }): Promise<void> {
+        const {
+          client,
+          value,
+          readerSchema,
+          processMessage,
+          writerSchema,
+          eventCount = crossLanguage ? 4 : 1,
+        } = settings;
+        if (!crossLanguage) {
```

Comment:
Should we generalize this to just be something like `alreadyEnqueued`?

Reply:
Good name, done.